clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java

```diff
@@ -358,12 +358,19 @@ private void process(final ShareAcknowledgeAsyncEvent event) {
      * consumer join the share group if it is not part of it yet, or send the updated subscription if
      * it is already a member.
      */
-    private void process(final ShareSubscriptionChangeEvent ignored) {
+    private void process(final ShareSubscriptionChangeEvent event) {
         if (!requestManagers.shareHeartbeatRequestManager.isPresent()) {
-            log.warn("Group membership manager not present when processing a subscribe event");
+            KafkaException error = new KafkaException("Group membership manager not present when processing a subscribe event");
+            event.future().completeExceptionally(error);
             return;
         }
+
+        if (subscriptions.subscribeToShareGroup(event.topics()))
+            metadata.requestUpdateForNewTopics();
```

Comment:
It seems this behavior is aligned with the async consumer, so my comment is not meant to block this PR. However, I'd like to discuss it for my own understanding 😄.

We only request to update the metadata cache here after the subscription changes. However, #17165 adds the condition if (subscriptionUpdated.compareAndSet(true, false) && state == MemberState.UNSUBSCRIBED), which prevents the consumer from rejoining the group if it’s already in the group.

This means the consumer could send a request with out-of-date assignments (since it doesn’t rejoin the group) but with updated metadata (since it subscribes to new topics). For example:
```java
consumer.subscribe(List.of("topic_a"));
...

consumer.subscribe(List.of("topic_b"));
```
If the topics in the second subscription weren't included in the previous subscription, this could cause a FETCH_SESSION_TOPIC_ID_ERROR until the consumer rejoins the group via a heartbeat.

Reply:
@kirktrue @lianetm @FrankYang0529 FYI

Reply:
Hi @chia7712, I tried to write a test case in `PlaintextConsumerSubscriptionTest`, but I can't reproduce it. Could you share your test case? Thank you.

```scala
  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumAndGroupProtocolNames)
  @MethodSource(Array("getTestQuorumAndGroupProtocolParametersAll"))
  def testSubscribeOtherTopicNotInPreviousSubscriptions(quorum: String, groupProtocol: String): Unit = {
    val initialAssignment = Set(new TopicPartition(topic, 0), new TopicPartition(topic, 1))
    val consumer = createConsumer()
    consumer.subscribe(List(topic).asJava)
    awaitAssignment(consumer, initialAssignment)
    assertDoesNotThrow(new Executable {
      override def execute(): Unit = consumer.poll(Duration.ofMillis(1000))
    })

    val otherTopic = "other"
    createTopic(otherTopic, 2, brokerCount)

    val otherAssignment = Set(new TopicPartition(otherTopic, 0), new TopicPartition(otherTopic, 1))
    consumer.subscribe(List(otherTopic).asJava)
    awaitAssignment(consumer, otherAssignment)
    assertDoesNotThrow(new Executable {
      override def execute(): Unit = consumer.poll(Duration.ofMillis(1000))
    })
  }
```

Reply:
@FrankYang0529 `FETCH_SESSION_TOPIC_ID_ERROR` is considered a recoverable error, so the consumer will update the metadata upon encountering it. Additionally, the heartbeat with the new subscription will trigger a group bump, leading to a rebalance. That's why your test case doesn't observe the error.